### PR TITLE
[chip-tool] Support complex types (as command argument or as writeable attribute)

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

`chip-tool` does not support complex types as command argument or as attribute when it comes to writing.

The first 4 commits are #13759 and #13726 and so they can be ignored while reviewing. The interesting part is the commit named: 
```
[chip-tool] Add complex attributes write supports
[chip-tool] Add support for complex command arguments
```

#### Change overview
 * Add complex types support

#### Testing
 I have verified it with a few different commands:
```
$ ./out/debug/standalone/chip-tool testcluster write list-int8u '[1, 2, 3, 4]' 0x12345 1 

$ ./out/debug/standalone/chip-tool testcluster write list-octet-string '["40", "10450000000000000000000000000000", "10", "042030456789"]' 0x12345 1

$ FOO_DATA='[{"nullableInt":null, "nullableString":null, "nullableStruct":null, "nullableList":null}, {"nullableInt":1, "nullableString":"Hello", "nullableOptionalInt": 2, "nullableStruct":{"a": 1, "b": false, "c": 1, "d":"000102", "e":"Hello", "f":1, "g":3.2, "h":43.4 }, "nullableList":[5, 4, 3, 2, 1]}]'
$ ./out/debug/standalone/chip-tool testcluster write list-nullables-and-optionals-struct ${FOO_DATA} 0X12345 1

$ ./out/debug/standalone/chip-tool testcluster simple-struct-echo-request '{"a":1, "b":false, "c":3, "d":"000102030002", "e":"Hello", "f":6, "g":7.6, "h":8.3333333}' 0x12345 1

$ ./out/debug/standalone/chip-tool thermostat set-weekly-schedule 1 1 1 '[1, 3, 4]' 0x12345 1
```

### CODE REVIEW

File: examples/chip-tool/commands/clusters/ComplexArgument.h [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        }
+
+        request = static_cast<T>(value.asUInt());
+        return CHIP_NO_ERROR;
+    }
+
+    template <typename T, std::enable_if_t<std::is_signed<T>::value, bool> = true>
+    static CHIP_ERROR Setup(const char * label, T & request, Json::Value value)
+    {
+        if (!value.isIntegral() || !chip::CanCastTo<T>(value.asLargestInt()))
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ComplexArgument.h:55]: Why is the unsigned case testing isNumeric() while this case tests isIntegral()?  Might be worth a comment.

File: examples/chip-tool/commands/common/Command.cpp [hunk 2]
```diff
@@ -479,6 +484,17 @@ size_t Command::AddArgument(const char * name, AddressWithInterface * out, uint8
     return AddArgumentToList(std::move(arg));
 }
 
+size_t Command::AddArgument(const char * name, ComplexArgument * value)
+{
+    Argument arg;
+    arg.type  = ArgumentType::Complex;
+    arg.name  = name;
+    arg.value = const_cast<void *>(reinterpret_cast<const void *>(value));
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:492]:
```suggestion
    arg.value = static_cast<void *>(value);
```

File: examples/chip-tool/commands/common/Command.cpp [hunk 3]
```diff
@@ -183,6 +183,11 @@ bool Command::InitArgument(size_t argIndex, char * argValue)
     Argument arg = mArgs.at(argIndex);
     switch (arg.type)
     {
+    case ArgumentType::Complex: {
+        ComplexArgument * complexArgument = reinterpret_cast<ComplexArgument *>(arg.value);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:187]:
```suggestion
        ComplexArgument * complexArgument = static_cast<ComplexArgument *>(arg.value);
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by gjc13 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]
Fast tracking, given tooling changes

File: examples/chip-tool/commands/clusters/ComplexArgument.h (hunk 1 above)
Comment (vivien-apple) [reply examples/chip-tool/commands/clusters/ComplexArgument.h]: I used `isNumeric` in the end. thanks !

**Review by andy31415 — APPROVED** [review]

**Pull request merged**